pkg/storage/internalstorage/storage.go

```diff
@@ -11,18 +13,64 @@ import (
 	"github.com/clusterpedia-io/clusterpedia/pkg/storage"
 )
 
+var mutex sync.Mutex
+
 type StorageFactory struct {
-	db *gorm.DB
+	db             *gorm.DB
+	AutoMigration  *bool
+	DivisionPolicy DivisionPolicy
+	Mapper         []ResourceMapper
+}
+
+func (s *StorageFactory) AutoMigrate() error {
+	return nil
 }
 
 func (s *StorageFactory) GetSupportedRequestVerbs() []string {
 	return []string{"get", "list"}
 }
 
 func (s *StorageFactory) NewResourceStorage(config *storage.ResourceStorageConfig) (storage.ResourceStorage, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var table string
+	if s.AutoMigration != nil && *s.AutoMigration {
+		switch s.DivisionPolicy {
+		case "", DivisionPolicyNone:
+			table = "resources"
+
+			if exist := s.db.Migrator().HasTable(table); !exist {
+				if err := s.db.AutoMigrate(&Resource{}); err != nil {
+					return nil, err
+				}
+			}
+		case DivisionPolicyGroupResource:
+			gvr := schema.GroupVersionResource{
+				Group:    config.StorageGroupResource.Group,
+				Version:  config.StorageVersion.Version,
+				Resource: config.StorageGroupResource.Resource,
+			}
+
+			table = GenerateTableFor(gvr)
+
+			if exist := s.db.Migrator().HasTable(table); !exist {
+				if err := s.db.AutoMigrate(&Resource{}); err != nil {
+					return nil, err
+				}
+
+				err := s.db.Migrator().RenameTable("resources", table)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
```

Comment:
what if the divisionpolicy is "custom" or other value?

Reply:
yes, I added default case in switch block.

Reply:
add default case may be better